DynamicStoragePiles/RestrictContainers.cs

```diff
@@ -0,0 +1,298 @@
+                var result = item.PrefabName() == allowedItem;
+                if (!result) {
+                    // Message player that item cannot be placed in container.
+                    var msg = $"{item.m_shared.m_name} cannnot be placed in {inventory.m_name}";
+                    Player.m_localPlayer?.Message(MessageHud.MessageType.Center, msg);
+                }
+                return result;
+            }
+            return true;
+        }
+
+        /// <summary>
+        ///     Check if container should be restricted and get the allowed item type for it
+        /// </summary>
+        /// <param name="containerName"></param>
+        /// <param name="allowedItem"></param>
+        /// <returns></returns>
+        public static bool IsRestrictedContainer(string containerName, out string allowedItem) {
+            if (!DynamicStoragePiles.ShouldRestrictItems) {
+                allowedItem = null;
+                return false;
+            }
+            return _allowedItemsByContainer.TryGetValue(containerName, out allowedItem);
+        }
+
+        /// <summary>
+        ///    Patch to trigger CanAddItem check when attempting to add item to inventory.
+        ///    If CanAddItem is false then the AddItem method in Valheim will be skipped
+        ///    and return false to indicate the item was not added.
+        /// </summary>
+        /// <param name="__instance"></param>
+        /// <param name="item"></param>
+        /// <param name="__result"></param>
+        /// <returns></returns>
+        [HarmonyPrefix]
+        [HarmonyPatch(nameof(Inventory.AddItem), new[] { typeof(ItemDrop.ItemData) })]
+        [HarmonyPriority(Priority.First)]
+        private static bool AddItemPrefix_1(Inventory __instance, ItemDrop.ItemData item, ref bool __result) {
+            if (!CanAddItem(__instance, item)) {
+                __result = false;
+                return __result;
+            }
+            return true;
+        }
+
+        /// <summary>
+        ///    Patch to trigger CanAddItem check when attempting to add item to inventory.
+        ///    If CanAddItem is false then the AddItem method in Valheim will be skipped
+        ///    and return false to indicate the item was not added.
+        /// </summary>
+        /// <param name="__instance"></param>
+        /// <param name="item"></param>
+        /// <param name="__result"></param>
+        /// <returns></returns>
+        [HarmonyPrefix]
+        [HarmonyPatch(nameof(Inventory.AddItem), new[] { typeof(ItemDrop.ItemData), typeof(int), typeof(int), typeof(int) })]
+        [HarmonyPriority(Priority.First)]
+        private static bool AddItemPrefix_2(Inventory __instance, ItemDrop.ItemData item, int amount, ref bool __result) {
+            if (!CanAddItem(__instance, item)) {
+                __result = false;
+                return __result;
+            }
+            return true;
+        }
```

Comment:
```suggestion
        /// <summary>
        ///    Patch to trigger CanAddItem check when attempting to add item to inventory.
        ///    If CanAddItem is false then the AddItem method in Valheim will be skipped
        ///    and return false to indicate the item was not added.
        /// </summary>
        /// <param name="__instance"></param>
        /// <param name="item"></param>
        /// <param name="__result"></param>
        /// <returns></returns>
        [HarmonyPrefix]
        [HarmonyPatch(nameof(Inventory.AddItem), new[] { typeof(ItemDrop.ItemData) })]
        [HarmonyPatch(nameof(Inventory.AddItem), new[] { typeof(ItemDrop.ItemData), typeof(int), typeof(int), typeof(int) })]
        [HarmonyPriority(Priority.First)]
        private static bool AddItemPrefix(Inventory __instance, ItemDrop.ItemData item, ref bool __result) {
            if (!CanAddItem(__instance, item)) {
                __result = false;
                return __result;
            }
            return true;
        }
```

The two patches can be merged together

Reply:
Wait, `typeof(Inventory)` is missing here as well. Did I miss a Harmony feature where the type isn't nessesary or was this case catched by a different patch?

Reply:
There is a Harmony feature where you can declare typeof(Inventory) on the class and it will apply to to all methods within the class so the code takes advantage of that since all the patch methods in the class are typeof(Inventory). Happy to change that if explicitly setting typeof for each method fits the conventions for DynamicStoragePiles better though.

Also thanks for letting me know about merging the patches together, I wasn't sure how Harmony would handle patching multiple methods based on partial signatures since I only want to patch 2 out of the 3 AddItem methods in the code. Now that I know they can be merged into one patch I'll do that though as it's a cleaner solution.

Reply:
Oh right, missed the HarmonyPatch at the class level. I rather prefer to explicitly put the type on the methods and not the class. The reason is readability IMO and it's easier to expand the class with a patch to a different class in the future.

Reply:
Yeah, I recently found that to be the case while working on another mod where I had to patch a variety of different classes to achieve a desired functionality. I've just added the explicit type to the AddItem patch methods and removed the typeof at the class level for RestricContianers.